### PR TITLE
Storage: Add glusterfs_volume module.

### PR DESCRIPTION
**Pull request by devyanikota**
##### SUMMARY
Adding GlusterFS Storage modules.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
Storage/glusterfs

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/devyani/work/gdeploy/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.13 (default, Dec  1 2017, 09:21:53) [GCC 6.4.1 20170727 (Red Hat 6.4.1-1)]
```

##### ADDITIONAL INFORMATION
Adding GlusterFS Storage modules.
This PR comprises of the glusterfs_volume module.
The glusterfs_volume module will set-up a Gluster cluster by creating volumes out of bricks.
Depending on the state selected, one can create/delete/start/stop and perform other operations
on the volume.
```

---
- hosts: 10.70.10.100
  remote_user: root
  roles:
    - volume-create

```
In the inventory file, group your gluster nodes in the 'host' group in a way that one of them
is the master node, which would be managing all the gluster operations and executions.

Signed-off-by: Devyani Kota <dkota@redhat.com>

### CODE REVIEW

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: glusterfs_volume
+short_description: Implementaion of GlusterFS
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: s/Implementaion/Implementation/

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+description:
+    - A module to set-up a GlusterFS cluster by
+      creating volumes out of bricks. The host group on which
+      playbooks will be written for this module should contain
+      only one hostname which should be a part of the cluster
+      and from which all the Gluster commands will be executed.
+      This will make sure the commands are only run once.
+version_added: "1.1"
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: s/1.1/2.5/

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+options:
+    action:
+        required: True
+        choices: [create, delete, start, stop, add-brick, remove-brick, replace-brick,
+                  rebalance, profile_state, bitrot, set,barrier]
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: space after comma.

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    state=bitrot
+    volume="{{ volname }}"
+    scrub="pause"
+'''
+
+import sys
+import re
+
+from collections import OrderedDict
+from ansible.module_utils.basic import *
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: Import specific.

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+        self.gluster_volume_ops()
+
+    def get_playbook_params(self, opt):
+        return self.module.params[opt]
+
+    def _validated_params(self, opt):
+        value = self.get_playbook_params(opt)
+        if value is None:
+            msg = "Please provide %s option in the playbook!" % opt
+            self.module.fail_json(rc=3,msg=msg)
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: space after ,

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+                option_str = self._validated_params('bitrot_daemon')
+
+        rc, output, err = self.call_gluster_cmd('volume', self.action,
+                                               volume, option_str, self.force)
+        self._get_output(rc, output, err)
+
+    def rebalance_volume(self):
+        state = self._validated_params('rebal_state')
+        if state not in ['start', 'stop', 'fix-layout']:
+            self.module.fail_json(rc=20,msg="Invalid state for rebalance action. \n" \
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: idem

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+def main():
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(required=True,
+                       choices=["absent", "present", "remove-brick", "started",
+                                "stopped", "replace-brick", "rebalance",
+                                "add-brick", "profile", "set", "barrier",
+                                "bitrot" ]),
+            hosts=dict(),
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: hosts is required as per documentation but does not reflect here.

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+        '''
+        If the bricks are not given in hostname:brickpath
+        format, hostnames are to be provided as
+        'hosts' in the playbook.
+        This method gets the bricks of all the hosts in the pool.
+        Expected to provide 'bricks' in the yml in the same
+        order as that of the hosts.
+        '''
+        if not self.hosts:
+            self.module.fail_json(rc=15,msg="Provide the bricks in hostname:brickpath " \
```
Comment (Akasurde) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: idem

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py (hunk 7 above)
Comment (devyanikota) [reply lib/ansible/modules/storage/glusterfs/glusterfs_volume.py]: Ack! added required flags for both 'hosts' and 'volname'. Thanks.

File: lib/ansible/modules/storage/glusterfs/glusterfs_volume.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+        required: False
+        description: Specifies the number of redundancy bricks. If
+        if not specified and the volume if of type disperse,
+        it's default value is computed so that it generates an
+        optimal configuration.
+    state:
+        required: True
+        choices: ["absent", "present", "remove-brick", "started",
+                  "stopped", "replace-brick", "rebalance",
+                  "add-brick", "profile", "set", "barrier", "bitrot" ]
```
Comment (sabose) [inline lib/ansible/modules/storage/glusterfs/glusterfs_volume.py:100]: state = barrier, bitrot, replace-brick, remove-brick etc seems like a wrong usage of state. For instance, there can be bitrot state as started/present but bitrot itself is not a state of a gluster volume.

**Comment by nerzhul** [comment]
@Akasurde can you close this ? we have a such module merged https://docs.ansible.com/ansible/latest/modules/gluster_volume_module.html

**Comment by devyanikota** [comment]
Closing this in favour of [PR#38269](https://github.com/ansible/ansible/pull/38269)

**Pull request closed without merging**